session: Prepare on one shard per node only

wprzytula opened the pull request:
## Overview

This PR comes with two main enhancements / optimisations:
1. **Prepare on every node, not on every shard.** This corresponds to #1290.
2. ~~**Wait only for one prepare attempt to succeed, not for all prepare attempts.** This is inspired by cpp-driver's logic.~~ (dropped)

Below, I describe both changes in detail.

## Prepare on every node, not on every shard

Just as other drivers do, now the Rust driver also prepares statements on a single connection to every node, not on a single connection to every shard. This brings performance benefits for both driver and cluster, as it lowers the overhead of handling duplicated prepare requests on both sides.

### Problem: unlucky random choice
As mentioned in https://github.com/scylladb/scylla-rust-driver/issues/1290#issuecomment-2740175939, we might be unlucky enough to end up with all randomly chosen connections being non-working. For example, the targeted shards might be overloaded.

### Solution
Fallback logic is introduced. Once we fail to prepare the statement on any node though the randomly chosen connections, the preparation is re-attempted **on all connections** (which is essentially how it used to work before this PR).

## ~~Wait only for one prepare attempt to succeed~~ (dropped)
~~Having taken a look at cpp-driver's handling of statement preparation, I noticed that the cpp-driver waits only for the first preparation to succeed. Preparation on all remaining nodes is done in the background, which decreases latency of the prepare operation from the driver user's PoV. To understand why, consider a node that is stuck/overloaded/temporarily unavailable. If we wait for prepare responses from all nodes, we are limited by the slowest response. This is not what ScyllaDB is designed for - it should be available and fast even if some nodes happen to be unavailable or overloaded.~~

~~I decided to implement this behaviour in the Rust driver. This is achieved by spawning a tokio worker task which prepares the statement on all nodes. It feeds all results, successes or failures, into a channel and this way signals the parent task. The parent task finishes early once it receives a successful response (a deserialized `PreparedStatement`). Meanwhile, the worker task keeps handling the remaining preparations in the background.~~

~~The change brings two main benefits:~~
    ~~1. reduces driver's latency upon statement preparation;~~
    ~~2. prevents the situation when one stuck node freezes the driver upon
       preparation - **there are no preparation timeouts!** (sic!).~~


## Tests

TODO. Waiting for #1246 to be merged. Then I'll add a new feature to the proxy to be aware of which shard was targeted by the intercepted frame. With that feature available, I'll write tests.

Fixes: #1290
## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [ ] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- ~~[ ] I have provided docstrings for the public items that I want to introduce.~~
- ~~[ ] I have adjusted the documentation in `./docs/source/`.~~
- [x] I added appropriate `Fixes:` annotations to PR description.

muzarski commented on scylla/src/client/session.rs:
```diff
@@ -1252,44 +1248,107 @@ impl Session {
-            first_ok.map_err(|first_attempt| PrepareError::AllAttemptsFailed { first_attempt })?;
-
-        // Validate prepared ids equality
-        for statement in results.flatten() {
-            if prepared.get_id() != statement.get_id() {
-                return Err(PrepareError::PreparedStatementIdsMismatch);
```
commit: `session: prepare() waits only for first success `

Notice that this optimization is a trade-off. We no longer check for statement id equality between the nodes - we cannot inform the user about it if it happens.

I'm not entirely sure if this is a good trade-off. While I like the optimization (because we are not blocked by one node that hangs), maybe a better idea would actually be to introduce (configurable) timeout and await all the preparations - then we could retain the statement verification logic.

IIUC, this would also allow us to ditch the complicated logic that spawns the separate tokio task (preparation_worker).

WDYT @Lorak-mmk ?

wprzytula replied:
The complex logic of the `preparation_worker` is not a problem to me. It's encapsulated, it's (I hope you agree) well documented, and it uses tokio abstractions to be more high-level.

What I'm not fully satisfied with is the overhead introduced by spawning a new tokio task upon each `prepare()`. However, I assume that `Session::prepare()` is called infrequently (perhaps only once per statement), so it should not hurt users.

Keep in mind that automatic/transparent preparation, which may occur more frequently than calls to `Session::prepare`, is not affected. By this I mean:
1. transparent preparation on a single connection upon `Session::query_*` with nonempty parameters,
2. automatic repreparation on a given connection upon `UNPREPARED` error received from the DB.

wprzytula replied:
> Notice that this optimization is a trade-off. We no longer check for statement id equality between the nodes - we cannot inform the user about it if it happens.

Well, we can still do that. It's easy to add such verification logic to the `preparation_worker` so that it logs a warning / an error-level message when it detects inequality.

muzarski replied:
> Should I add such checks?

I'll let you decide. If it's fairly easy to implement, and won't drastically increase the complexity of the `preparation_workter`, I'd say why not. I think error-level message is appropriate - the id mismatch indicates a serious bug on the server side. UUIC, the id is just a hash of the statement contents, so it definitely should be the same on different nodes.

muzarski commented on scylla/src/client/session.rs:
```diff
@@ -1252,44 +1248,107 @@ impl Session {
+        }
+
+        tokio::task::spawn(preparation_worker(
+            cluster_state.clone(),
```
commit: `session: prepare() waits only for first success `

nit: `Arc::clone()`

muzarski commented on scylla/src/client/session.rs:
```diff
@@ -1304,62 +1304,80 @@ impl Session {
-                Err(attempt_error) => {
-                    if first_error.is_none() {
-                        first_error = Some(attempt_error);
+            tokio::task::spawn(preparation_worker(
+                cluster_state.clone(),
```
commit: `session: prepare: extract prepare_on_all()`

ditto: `Arc::clone()`

muzarski reviewed (changes requested):
Looks really good. I just wonder - maybe a better idea than spawning a separate task is to just introduce configurable timeout for each attempt (see my other comment).

wprzytula commented on scylla/src/cluster/state.rs:
```diff
@@ -344,6 +344,28 @@ impl ClusterState {
+
+        // First we try to find the first working pool of connections.
+        // If none is found, return error.
+        let first_working_pool = peers_iter
+            .by_ref()
+            .map(|node| node.get_random_connection())
+            .find_or_first(Result::is_ok)
+            .expect("impossible: known_peers was asserted to be nonempty")?;
+
+        let remaining_pools_iter = peers_iter.flat_map(|node| node.get_random_connection());
```
> But... As I mentioned, it looks like the errors in `Node::get_random_connection()` are handled per-pool, and not per-connection. So this looks fine to me - just wanted to point it out.

Yep. `ConnectionPool::random_connection()` fails iff there's no working connection in the whole pool.

wprzytula commented:
v.1.0.2: rebased on main.

wprzytula commented:
v1.1:
- replaced internal calls to `Session::prepare()` with `Session::prepare_nongeneric()`;
- added verification of prepared id equality in the worker task.

wprzytula commented:
v1.1: rebased on main. Incorporated changes from #1329.

muzarski commented on scylla/src/client/session.rs:
```diff
@@ -1253,41 +1251,91 @@ impl Session {
+
+                    // This is the first preparation that succeeded.
+                    // Let's return the PreparedStatement.
+                    // Preparation on other nodes will continue in the background tokio task.
+                    prepared.set_partitioner_name(
+                        Session::extract_partitioner_name(&prepared, cluster_state)
+                            .and_then(PartitionerName::from_str)
+                            .unwrap_or_default(),
+                    );
```
nit: Outdated comment

muzarski commented on scylla/src/client/session.rs:
```diff
@@ -1253,41 +1251,91 @@ impl Session {
+        {
+            let on_all_nodes_result = prepare_on_all(statement, &cluster_state, false).await;
+            if let Ok(prepared) = on_all_nodes_result {
+                // We succeeded in preparing the statement on at least one node. We're done; at the same time,
+                // the background tokio task attempts preparation on remaining nodes.
```
ditto: outdated comment

muzarski commented on scylla/src/client/session.rs:
```diff
@@ -1259,15 +1259,9 @@ impl Session {
             statement: &Statement,
             cluster_state: &ClusterState,
-            on_all_shards: bool,
+            working_connections: &mut (dyn Iterator<Item = Arc<Connection>> + Send),
```
I agree that `dyn Iterator` is cleaner.

Lorak-mmk replied:
Commit: session: prepare_on_all() uses dyn Iterator

Ah, I didn't know about existance of this commit when writing my comment about dyn Iterator.

wprzytula commented:
v2.0.1: fixed outdated comments.

Lorak-mmk commented on scylla/src/cluster/state.rs:
```diff
@@ -344,6 +344,28 @@ impl ClusterState {
         // is nonempty, too.
     }
 
+    /// Returns nonempty iterator of working connections to all nodes.
+    pub(crate) fn iter_working_connections_to_nodes(
+        &self,
+    ) -> Result<impl Iterator<Item = Arc<Connection>> + '_, ConnectionPoolError> {
+        // The returned iterator is nonempty by nonemptiness invariant of `self.known_peers`.
+        assert!(!self.known_peers.is_empty());
+        let mut peers_iter = self.known_peers.values();
+
+        // First we try to find the first working pool of connections.
+        // If none is found, return error.
+        let first_working_pool = peers_iter
+            .by_ref()
+            .map(|node| node.get_random_connection())
+            .find_or_first(Result::is_ok)
+            .expect("impossible: known_peers was asserted to be nonempty")?;
+
+        let remaining_pools_iter = peers_iter.flat_map(|node| node.get_random_connection());
+
+        Ok(std::iter::once(first_working_pool).chain(remaining_pools_iter))
+        // The returned iterator is nonempty, because it returns at least `first_working_pool`.
```
This iterator usage is too complicated for me to read.
I see that we return Result, with the Ok variant being an iterator of connections (no results of connections).
I don't see where errors from `node.get_random_connection` in `remaining_pools_iter` are being handled.

Please add either explicit types, or comments, so that this function is easier to understand.

Lorak-mmk commented on scylla/src/client/session.rs:
```diff
@@ -1199,6 +1199,14 @@ impl Session {
     /// Prepares a statement on the server side and returns a prepared statement,
     /// which can later be used to perform more efficient requests.
     ///
+    /// The statement is prepared on all nodes. This function finishes once all nodes respond
+    /// with either success or an error.
+    // TODO: There are no timeouts here. So, just one stuck node freezes the driver here, potentially indefinitely long.
+    // Also, what the driver requires to get from the cluster is the prepared statement metadata.
+    // It suffices that it gets only one copy of it, just from one success response. Therefore, it's a possible
+    // optimisation that the function only waits for one preparation to finish successfully, and then it returns.
+    // For it to be done, other preparations must continue in the background, on a separate tokio task.
+    ///
```
The described idea is imo not a TODO, because it is not obvious it is something we should do. It should be an issue instead.

wprzytula replied:
Why shouldn't it be a TODO? Issues have the problem that they are not immediately visible when someone is reading the code. OTOH, such a TODO catches attention quickly and makes the reader aware of the described problem and considered solutions.

Lorak-mmk commented on scylla/src/client/session.rs:
```diff
@@ -1253,41 +1251,91 @@ impl Session {
+                    Ok(prepared)
+                }
+                Err(err) => Err(PrepareError::AllAttemptsFailed { first_attempt: err }),
+            }
```
I don't see why this is a nested function. Please make it just a private top-level function, like prepare_nongeneric.

Lorak-mmk commented on scylla/src/client/session.rs:
```diff
@@ -1253,41 +1251,91 @@ impl Session {
+                itertools::Either::Left(cluster_state.iter_working_connections_to_shards()?)
+            } else {
+                itertools::Either::Right(cluster_state.iter_working_connections_to_nodes()?)
+            };
```
Instead of having weird boolean parameter you could accept dyn Iterator

Lorak-mmk commented on scylla/src/client/session.rs:
```diff
@@ -1253,41 +1251,91 @@ impl Session {
+            let mut raw_prepared_statements_results_iter =
+                raw_prepared_statements_results.into_iter();
+
+            // Safety: We pass a nonempty iterator, so there will be at least one result.
+            let first_ok_or_error = raw_prepared_statements_results_iter
+                .by_ref()
+                .find_or_last(|res| res.is_ok())
+                .unwrap(); // Safety: there is at least one connection.
```
Commit: "session: prepare: introduce fallback to all shards "

Previously `find_or_first` was used, now you changed it to `find_or_last`. Why?

wprzytula commented:
v2.1:
- squashed the `&mut dyn Iterator` commit into the fallback commit,
- extracted more commits to make review smoother,
- improved readability of the `iter_working_connections*()` methods,
- brought back the, accidentally removed, logic that collects prepared tracing IDs from prepared statements on all connections.

wprzytula commented on scylla/src/cluster/state.rs:
```diff
@@ -324,22 +324,35 @@ impl ClusterState {
-            .flatten();
-
-        Ok(first_working_pool.into_iter().chain(remaining_pools_iter))
+        let first_working_pool_or_error: Result<Vec<Arc<Connection>>, ConnectionPoolError> =
+            connection_pool_per_node_iter
+                .by_ref()
+                .find_or_first(Result::is_ok)
+                .expect("impossible: known_peers was asserted to be nonempty");
+
+        // We have:
+        // 1. either consumed the whole iterator without success and got the first error,
+        //    in which case we propagate it;
+        // 2. or found the first working pool of connections.
+        let first_working_pool: Vec<Arc<Connection>> = first_working_pool_or_error?;
+
+        // We retrieve connection pools for remaining nodes (those that are left in the iterator
+        // once the first working pool has been found).
+        let remaining_pools_iter = connection_pool_per_node_iter;
+        // Pools are flattened, so now we have `impl Iterator<Item = Result<Arc<Connection>, ConnectionPoolError>>`.
+        let remaining_connections_iter = remaining_pools_iter.flatten_ok();
+        // Errors (non-working pools) are filtered out.
+        let remaining_working_connections_iter = remaining_connections_iter.flatten();
```
> There I was confused again, and I had to check in documentation that Result implements IntoIterator, returning iterator that will have 1 element for Ok, and 0 for Err.

TBH, I thought this is a widely known and used property, similat to how `Option` implements `IntoIterator`.



> I understand that this is correct, but I also think this is a bit confusing. Maybe `.filter_map(Result::ok)` would be more explicit and obvious?

No problem, I can write it this way.

Lorak-mmk replied:
> TBH, I thought this is a widely known and used property, similarly to how Option implements IntoIterator.

I knew about Option. I did not suspect Result does that too, because for Result such impl ignores errors in imo too silent way,

wprzytula commented:
v2.1.1: adressed @Lorak-mmk comments.

wprzytula commented:
Rebased on main.
